File: gradle/packages/cli-function/src/main/kotlin/com/prog/cli/function/Details.kt
package com.prog.cli.function

import com.prog.hello.world.printDate
import com.prog.translator.translate

object Details {
    /**
     * # `main`
     *
     * This function is entry point of the application.
     *
     * __Arguments:__
     *
     * * `args: Array<String>` - Command line arguments
     */
    @JvmStatic
    fun main(args: Array<String>) {
        println("===================================================")
        printUserDetails(args)
        println("===================================================")
    }

    /**
     * # `getUserName`
     *
     * This function prompts the user to enter their name and returns the input
     *
     * __Arguments:__
     *
     * * `args: Array<String>` - Command line arguments
     *
     * __Return:__
     *
     * * `String` - The user's name
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.cli.function.Details.getUserName
     * val name = getUserName(arrayOf("John", "Doe"))
     * println(name)
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "John Doe"
     * ```
     */
    private fun getUserName(args: Array<String>): String {
        return if (args.isEmpty()) {
            print("Please enter your name : ")
            readLine()!!.trim()
        } else {
            args.joinToString(" ")
        }
    }

    /**
     * # `getUserLanguage`
     *
     * This function prompts the user to enter their language and returns the
     * input.
     *
     * __Return:__
     *
     * `String` - The user's language
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.cli.function.Details.getUserLanguage
     * val language = getUserLanguage()
     * println(language)
     * ```
     */
    private fun getUserLanguage(): String {
        print("Please enter your language : ")
        return readLine()!!.trim()
    }

    /**
     * # `printUserDetails`
     *
     * This function prints the user's name and the current date. It also prints
     * the translated message based on the user's language. The user's name and
     * language are obtained from the command line arguments and user input using
     * the `getUserName` and `getUserLanguage` functions.
     *
     * __Arguments:__
     *
     * * `Array<String>` - Command line arguments
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.cli.function.Details.printUserDetails
     * printUserDetails(arrayOf("John", "Doe"))
     * ```
     */
    fun printUserDetails(args: Array<String>) {
        val name = getUserName(args)
        val language = getUserLanguage()
        println(translate(language, name))
        println(printDate())
    }
}

File: haskell/src/CLI/Function/Details.hs
module CLI.Function.Details (printUserDetails) where

import System.IO (hFlush, stdout)
import Translator (translate)

import Hello.World (printDate)

-- |
-- = @getUserName@
--
-- The function @getUserName@ is used to get the user name.
--
-- __Arguments:__
--
-- * @args: [String]@ - The command line arguments.
--
-- __Return:__
--
-- * @IO String@ - The user name.
getUserName :: [String] -> IO String
getUserName args =
    if null args
        then do
            putStr "Please enter your name : "
            hFlush stdout
            getLine
        else
            return (unwords args)

-- |
-- = @getUserLanguage@
--
-- The function @getUserLanguage@ is used to get the user language.
--
-- __Return:__
--
-- * @IO String@ - The user language.
getUserLanguage :: IO String
getUserLanguage = do
    putStr "Please enter your language : "
    hFlush stdout
    getLine

-- |
-- = @printUserDetails@
--
-- The function @printUserDetails@ is used to print the user details.
--
-- __Arguments:__
--
-- * @args: [String]@ - The command line arguments.
--
-- __Return:__
--
-- * @IO ()@ - The user details.
printUserDetails :: [String] -> IO ()
printUserDetails args = do
    name <- getUserName args
    language <- getUserLanguage
    putStrLn $ translate language name
    putStrLn =<< printDate
